_posts/2021-06-29-fed-talk-s01e04-react-components.md

```diff
@@ -0,0 +1,545 @@
+
+> The Effect Hook, useEffect, adds the ability to perform side effects from a function component. It serves the same purpose as componentDidMount, componentDidUpdate, and componentWillUnmount in React classes, but unified into a single API.
+
+```tsx
+function NameField() {
+  React.useEffect(() => {
+    console.log("On Mount")
+  }, [])
+
+  console.log('Before Render')
+  return (
+    <input type="text" value="" placeholder="Enter your name" />
+  )
+}
+```
+
+**Note**, for the component's mount phase to be triggered once, its most important for the dependency array to be empty as React treats this use case in a special by, where by it will ensure it only gets invoked once for the lifetime of the component being mounted.
+
+Order of events:
+1. `Before Render` will be logged
+2. `<input />` field would render
+3. `On Mount` will be logged
+
+Lets explore what will happen if your component's mount phase cause a side effect.
+
+```tsx
+function NameField() {
+  const [currentState, setCurrentState] = React.useState("beforeMounted");
+  React.useEffect(() => {
+    console.log("On Mount");
+    setCurrentState("mounted");
+  }, [])
+
+  console.log("Before Render: ", currentState)
+  return (
+    <input type="text" value="" placeholder="Enter your name" />
+  )
+}
+```
+
+Order of events:
+1. `Before Render: beforeMounted` will be logged
+2. `<input />` field would render
+3. `On Mount` will be logged
+4. Component State gets updated
+5. `Before Render: mounted` will be logged
+6. `<input />` field re-renders
+
+### Component Un-Mount Phase
+
+Like the Component Mount phase, React conveniently offers an Un-Mount phase which you can piggy-back from the the Mount implementation. This hook is underrated and sometimes forgotten. Its purpose is to cleanup any of the components external factors prior to React un-mounting the component from its tree. A typical use case would be to abort asynchronous operations, or to alter application state etc.
+
+Lets take a look what this looks like:
+
+```tsx
+function NameField() {
+  const [currentState, setCurrentState] = React.useState("beforeMounted");
+  React.useEffect(() => {
+    console.log("On Mount")
+    setCurrentState("mounted");
+
+    return () => {
+      console.log("Un Mounting")
+      setCurrentState("beforeUnMounting")
+    }
+  }, [])
+
+  console.log("Before Render: ", currentState)
+  return (
+    <input type="text" value="" placeholder="Enter your name" />
+  )
+}
+```
+
+Order of events:
+1. `Before Render: beforeMounted` will be logged
+2. `<input />` field would render
+3. `On Mount` will be logged
+4. Component State gets updated
+5. `Before Render: mounted` will be logged
+6. `<input />` field re-renders
+7. ... something un-mounts the component
+8. `Un Mounting` will be logged
+
+**Note** even though we `setCurrentState()` in the un-mount function, it's side affect is disregarded as the next operation is to destroy that function anyway.
+
```

Comment:
As far as I know, you can't access `useState` variables in the un-mount hook but if you needed to, I found this useful blog post: https://www.timveletta.com/blog/2020-07-14-accessing-react-state-in-your-component-cleanup-with-hooks/

Reply:
heyyyy **Tim Velettas** website!! hahaha _"I found this useful blog post"_. 
My statement was more around the fact that you `can` access state, but if you try to update internal component state in the unmount phase it'll do nothing.